File: kernel/cpu/x86_64/src/mmu.c

// SPDX-License-Identifier: MIT

#include "cpu/mmu.h"

#include "assertions.h"
#include "cpu/x86_cpuid.h"
#include "cpu/x86_cr.h"
#include "interrupt.h"
#include "isr_ctx.h"

_Static_assert(MEMMAP_PAGE_SIZE == MMU_PAGE_SIZE, "MEMMAP_PAGE_SIZE must equal MMU_PAGE_SIZE");



// Virtual address offset used for HHDM.
size_t mmu_hhdm_vaddr;
// Virtual address offset of the higher half.
size_t mmu_high_vaddr;
// Size of a "half".
size_t mmu_half_size;
// How large to make the HHDM, rounded up to pages.
size_t mmu_hhdm_size;
// Number of page table levels.
int    mmu_levels;
// SMAP is supported.
bool   smap_support;
// Process ID bits are supported.
bool   pcid_support;



// Load a word from physical memory.
static inline size_t pmem_load(size_t paddr) {
    assert_dev_drop(paddr < mmu_hhdm_size);
    return *(size_t volatile *)(paddr + mmu_hhdm_vaddr);
}

// Store a word to physical memory.
static inline void pmem_store(size_t paddr, size_t data) {
    assert_dev_drop(paddr < mmu_hhdm_size);
    *(size_t volatile *)(paddr + mmu_hhdm_vaddr) = data;
}



// MMU-specific init code.
void mmu_early_init() {
    // Enable NXE feature.
    msr_efer_t efer;
    efer.val = msr_read(MSR_EFER);
    efer.nxe = 1;
    msr_write(MSR_EFER, efer.val);

    // Check for SMAP support.
    smap_support = false; // cpuid(0x07).ebx & (1 << 20);
    pcid_support = false; // cpuid(0x17).ecx & (1 << 17);

    // Enable PCIDE and SMAP if supported.
    x86_cr4_t cr4;
    asm volatile("mov %0, cr4" : "=r"(cr4));
    cr4.pcide = pcid_support;
    cr4.smap  = smap_support;
    asm volatile("mov cr4, %0" ::"r"(cr4));

    // TODO: Detect number of paging levels supported by the CPU.
    mmu_levels     = 4;
    mmu_half_size  = 1llu << (11 + 9 * mmu_levels);
    mmu_high_vaddr = -mmu_half_size;

    if (smap_support) {
        // The kernel shall, by default, not access user memory.
        asm volatile("clac" ::: "memory");
    }
}

// MMU-specific init code.
void mmu_init() {
}



// Read a PTE from the page table.
mmu_pte_t mmu_read_pte(size_t pte_paddr) {
    return (mmu_pte_t){.val = pmem_load(pte_paddr)};
}

// Write a PTE to the page table.
void mmu_write_pte(size_t pte_paddr, mmu_pte_t pte) {
    pmem_store(pte_paddr, pte.val);
}



// Swap in memory protections for the given context.
void memprotect_swap_from_isr() {
    memprotect_swap(isr_ctx_get()->mpu_ctx);
}

// Swap in memory protections for a given context.
void memprotect_swap(mpu_ctx_t *mpu) {
    mpu           = mpu ?: &mpu_global_ctx;
    bool      ie  = irq_disable();
    x86_cr3_t cr3 = {
        .root_ppn = mpu->root_ppn,
    };
    asm("mov %%cr3, %0" ::"r"(cr3));
    isr_ctx_get()->mpu_ctx = mpu;
    irq_enable_if(ie);
}

File: kernel/cpu/x86_64/src/mmu.rs
// SPDX-FileCopyrightText: 2025 Julian Scheffers <julian@scheffers.net>
// SPDX-FileType: SOURCE
// SPDX-License-Identifier: MIT

use core::arch::asm;

use crate::mem::{
    pmm::PPN,
    vmm::mmu::{ASID_BITS, PAGING_LEVELS, PTE},
};

#[derive(Debug, Clone, Copy)]
/// Data type that can store a packed page table entry.
pub struct PackedPTE(usize);

impl PackedPTE {
    /// An invalid PTE with no special data in it.
    pub const INVALID: PackedPTE = PackedPTE(0);

    /// Unpack this PTE.
    pub fn unpack(self, level: u8) -> PTE {
        todo!();
    }
}

impl PTE {
    /// Pack this PTE.
    pub fn pack(self) -> PackedPTE {
        todo!();
    }
}

/// Maximum possible value of ASID.
pub const ASID_MAX: usize = 0xffff;

/// Number of virtual address bits per page table level.
pub const BITS_PER_LEVEL: usize = 9;

/// Initialize and detect capabilities of the MMU, given the constructed page table.
pub unsafe fn init(root_ppn: PPN) {
    todo!();
}

#[inline(always)]
/// Switch page table and address space ID.
pub unsafe fn set_page_table(root_ppn: PPN, asid: usize) {
    todo!();
}

#[inline(always)]
/// Perform a fence of virtual memory.
pub fn vmem_fence(vaddr: Option<usize>, asid: Option<usize>) {
    todo!();
}
